### PR TITLE
Multistage dockerfile

### LINKED ISSUE #950: Versioneer not working in docker container
Due to the way we are constructing the docker container (not adding the `.git` directory) versioneer is unable to calculate the correct version (`0+unknown`).

We should figure out a way, maybe with multi-stage builds, to correctly retain the version info.

### PR DESCRIPTION
**Pull request by Dissonant-Tech**
# Description

Fixes #950

Updates the Dockerfile to leverage multi-stage build and ensure that the
runtime contains only the dependencies required to run OpsDroid.

The final image was updated to run as an unprivileged user.

The concepts used to build this new version of the Docker image mostly
come from Bitnami's recommendations:
https://engineering.bitnami.com/articles/best-practices-writing-a-dockerfile.html

The wheel unversal flag was removed since the project does not have C
extensions, nor supports Python 2 (see
https://packaging.python.org/guides/distributing-packages-using-setuptools/#wheels
for reference).

## Status
**UNDER DEVELOPMENT**

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Building and running the image locally.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation (if applicable)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by jacobtomlinson** [comment]
This looks a lot like  #1794 (but not exactly?). Is this intended to supersede it?

**Comment by Dissonant-Tech** [comment]
> This looks a lot like #1794 (but not exactly?). Is this intended to supersede it?

I made it a draft PR for now to have the CI run on a branch that contains the changes suggested in the conversations for #1794. 
Will set to "ready for review" in case @rgreinho does not have the time to add the suggested changes..

**Comment by jacobtomlinson** [comment]
Great thanks @Dissonant-Tech. It's been a few weeks since #1794 has been worked on. 

As you're a first time contributor I think I will need to manually approve each CI run whenever you push commits. Feel free to open a small superficial PR for me to merge quickly so that you get CI run privileges.

I also notice you haven't applied this suggestion in this PR https://github.com/opsdroid/opsdroid/pull/1794#discussion_r679640785

**Comment by Dissonant-Tech** [comment]
> Great thanks @Dissonant-Tech. It's been a few weeks since #1794 has been worked on.
> 
> As you're a first time contributor I think I will need to manually approve each CI run whenever you push commits. Feel free to open a small superficial PR for me to merge quickly so that you get CI run privileges.
> 
> I also notice you haven't applied this suggestion in this PR [#1794 (comment)](https://github.com/opsdroid/opsdroid/pull/1794#discussion_r679640785)

That's my own suggestion too! Must have gotten lost in the merge :man_facepalming:. Thanks for catching that.

I think this should be the last workflow you need to approve. Was just fixing the `docker-min` build errors which was the only failing CI job. If it does fail then I'll do the superficial PR :+1:

File: Dockerfile [hunk 1]
```diff
@@ -1,18 +1,66 @@
...
+
+ARG EXTRAS=[all,connector_matrix_e2e]
+ENV DEPS_DIR=/usr/src/app/deps
+
+# Copy the pre-built dependencies.
+COPY --from=builder ${DEPS_DIR}/*.whl ${DEPS_DIR}/
+
+# Install Opsdroid using only pre-built dependencies.
+RUN apk add --no-cache \
+    git \
```
Comment (Cadair) [inline Dockerfile:54]: As I said on the other PR I am pretty sure we need olm at runtime here.

Comment (Dissonant-Tech) [reply Dockerfile:54]: Done :+1: .

**Comment by Dissonant-Tech** [comment]
Added `olm` to the runtime image. Also can confirm that versioneer is fixed (I think) now that git is installed during the build stage 

Screenshot from 2021-08-17 12-53-19


Currently the docker-min test is still failing. From what I can gather it's due to permission errors when the tests run `docker exec opsdroid-container sh -c 'apk add --no-cache curl && curl -sSf http://localhost:8080/ || exit 1'` as the docker image now uses a non-root user and does not have permissions to run `apk`. Should be fixed by adding  `--user root` to the previously mentioned command, will try this soon.

**Comment by ajitdsa** [comment]
Is there anything keeping this from being merged?

File: Dockerfile [hunk 2]
```diff
@@ -1,18 +1,67 @@
-FROM python:3.8-alpine
+FROM python:3.9.6-alpine3.14 as builder
 LABEL maintainer="Jacob Tomlinson <jacob@tomlinson.email>"
+LABEL maintainer="Rémy Greinhofer <remy.greinhofer@gmail.com>"
```
Comment (jacobtomlinson) [inline Dockerfile:3]:
```suggestion
```

**Comment by jacobtomlinson** [comment]
I don't think so. I had missed that things were fixed here (force pushing makes reviewing on GitHub much harder).

**Comment by ajitdsa** [comment]
Awesome, thanks so much!

**Pull request merged**